File: _archived/OGS-Client/js/core/sgf-parser.js
/**
 * SGF Parser - Direct port from SGFPlayer Swift code
 * Lightweight SGF AST (MVP: flattens to the main line; ignores variations)
 */

class SGFError extends Error {
    constructor(message) {
        super(message);
        this.name = 'SGFError';
    }
}

class SGFNode {
    constructor(props = {}) {
        this.props = props;
    }
}

class SGFTree {
    constructor(nodes = []) {
        this.nodes = nodes;
    }
}

class SGFParser {
    static parse(text) {
        // Normalize line endings
        const s = text.replace(/\r/g, '');
        let i = 0;

        function peek() {
            return i < s.length ? s[i] : null;
        }

        function advance() {
            if (i < s.length) i++;
        }

        function skipWhitespace() {
            while (peek() && /\s/.test(peek())) {
                advance();
            }
        }

        const nodes = [];
        skipWhitespace();

        if (peek() !== '(') {
            throw new SGFError("Missing '(' at start");
        }

        consumeSubtree(); // parse the outer tree, collecting the main line nodes
        return new SGFTree(nodes);

        function consumeSubtree() {
            if (peek() !== '(') {
                throw new SGFError("Expected '('");
            }
            advance(); // '('
            skipWhitespace();

            while (peek()) {
                const c = peek();
                if (c === ';') {
                    advance();
                    nodes.push(parseNode());
                    skipWhitespace();
                } else if (c === '(') {
                    skipSubtree(); // skip variation
                    skipWhitespace();
                } else if (c === ')') {
                    advance();
                    break;
                } else {
                    advance(); // tolerate junk
                }
            }
        }

        function parseNode() {
            const props = {};
            skipWhitespace();

            while (peek() && /[a-zA-Z]/.test(peek())) {
                const key = parseIdent();
                const values = [];
                skipWhitespace();

                while (peek() === '[') {
                    values.push(parseValue());
                    skipWhitespace();
                }

                if (values.length > 0) {
                    if (!props[key]) props[key] = [];
                    props[key].push(...values);
                }
                skipWhitespace();
            }

            return new SGFNode(props);
        }

        function parseIdent() {
            const start = i;
            while (peek() && /[a-zA-Z]/.test(peek())) {
                advance();
            }
            return s.substring(start, i).toUpperCase();
        }

        function parseValue() {
            if (peek() !== '[') {
                throw new SGFError("Expected '['");
            }
            advance(); // '['

            let out = '';
            while (peek()) {
                const c = peek();
                advance();
                if (c === '\\') {
                    if (peek()) {
                        out += peek();
                        advance();
                    }
                } else if (c === ']') {
                    break;
                } else {
                    out += c;
                }
            }
            return out;
        }

        function skipSubtree() {
            if (peek() !== '(') return;
            advance();
            let depth = 0;

            while (peek()) {
                const c = peek();
                advance();
                if (c === '(') {
                    depth++;
                } else if (c === ')') {
                    if (depth === 0) {
                        break;
                    } else {
                        depth--;
                    }
                }
            }
        }
    }
}

// Stone types
const Stone = {
    BLACK: 'black',
    WHITE: 'white'
};

class SGFGame {
    constructor() {
        this.boardSize = 19;
        this.info = {
            event: null,
            playerBlack: null,
            playerWhite: null,
            result: null,
            date: null
        };
        this.setup = []; // [{stone: Stone, x: number, y: number}]
        this.moves = []; // [{stone: Stone, coord: {x: number, y: number} | null}] (null = pass)
    }

    static from(tree) {
        const game = new SGFGame();

        for (const node of tree.nodes) {
            for (const [k, vals] of Object.entries(node.props)) {
                switch (k) {
                    case 'SZ':
                        if (vals[0]) {
                            const v = vals[0];
                            if (v.includes(':')) {
                                const [left, right] = v.split(':').map(s => parseInt(s) || 19);
                                game.boardSize = Math.max(2, Math.min(25, Math.min(left, right)));
                            } else {
                                const sz = parseInt(v) || 19;
                                game.boardSize = Math.max(2, Math.min(25, sz));
                            }
                        }
                        break;
                    case 'EV':
                        game.info.event = vals[0];
                        break;
                    case 'PB':
                        game.info.playerBlack = vals[0];
                        break;
                    case 'PW':
                        game.info.playerWhite = vals[0];
                        break;
                    case 'RE':
                        game.info.result = vals[0];
                        break;
                    case 'DT':
                        game.info.date = vals[0];
                        break;
                    case 'AB':
                        for (const v of vals) {
                            const coord = parseCoord(v);
                            if (coord) {
                                game.setup.push({stone: Stone.BLACK, x: coord.x, y: coord.y});
                            }
                        }
                        break;
                    case 'AW':
                        for (const v of vals) {
                            const coord = parseCoord(v);
                            if (coord) {
                                game.setup.push({stone: Stone.WHITE, x: coord.x, y: coord.y});
                            }
                        }
                        break;
                    case 'B':
                        const bCoord = parseCoord(vals[0] || '');
                        game.moves.push({stone: Stone.BLACK, coord: bCoord});
                        break;
                    case 'W':
                        const wCoord = parseCoord(vals[0] || '');
                        game.moves.push({stone: Stone.WHITE, coord: wCoord});
                        break;
                }
            }
        }

        return game;
    }
}

// SGF coords: two lowercase letters; 'aa' => (0,0). Empty => pass.
function parseCoord(s) {
    if (!s) return null; // pass
    if (s.length !== 2) return null;

    const x = s.charCodeAt(0) - 97; // 'a' = 97
    const y = s.charCodeAt(1) - 97;

    if (x < 0 || y < 0) return null;
    return {x, y};
}

// Export for web usage
window.SGFParser = SGFParser;
window.SGFGame = SGFGame;
window.SGFError = SGFError;
window.Stone = Stone;

File: Old/Sources/GoCore/Domain/SGFParser.swift
// MARK: - SGF Parser
// Extracted from SGFKit.swift for reusability across all Go applications

import Foundation

/// Errors that can occur during SGF parsing
public enum SGFParsingError: Error, LocalizedError {
    case parseError(String)
    case invalidFormat(String)
    case unsupportedFeature(String)

    public var errorDescription: String? {
        switch self {
        case .parseError(let message):
            return "SGF Parse Error: \(message)"
        case .invalidFormat(let message):
            return "Invalid SGF Format: \(message)"
        case .unsupportedFeature(let message):
            return "Unsupported SGF Feature: \(message)"
        }
    }
}

/// Lightweight SGF AST - represents the parsed structure
public struct SGFTree {
    public let nodes: [SGFNode]

    public init(nodes: [SGFNode]) {
        self.nodes = nodes
    }
}

/// Individual node in SGF tree
public struct SGFNode {
    public var properties: [String: [String]]

    public init(properties: [String: [String]] = [:]) {
        self.properties = properties
    }

    /// Get first value for a property
    public func getValue(_ key: String) -> String? {
        return properties[key]?.first
    }

    /// Get all values for a property
    public func getValues(_ key: String) -> [String] {
        return properties[key] ?? []
    }

    /// Check if property exists
    public func hasProperty(_ key: String) -> Bool {
        return properties[key] != nil
    }
}

/// Main SGF parser
public enum SGFParser {
    /// Parse SGF text into a game tree
    public static func parseTree(_ text: String) throws -> SGFTree {
        let normalizedText = text.replacingOccurrences(of: "\r", with: "")
        var index = normalizedText.startIndex

        func peek() -> Character? {
            return index < normalizedText.endIndex ? normalizedText[index] : nil
        }

        func advance() {
            if index < normalizedText.endIndex {
                index = normalizedText.index(after: index)
            }
        }

        func skipWhitespace() {
            while let char = peek(), char.isWhitespace {
                advance()
            }
        }

        var nodes: [SGFNode] = []

        func parseSubtree() throws {
            guard peek() == "(" else {
                throw SGFParsingError.parseError("Expected '(' at start of subtree")
            }
            advance() // Skip '('
            skipWhitespace()

            while let char = peek() {
                if char == ";" {
                    advance() // Skip ';'
                    nodes.append(try parseNode())
                    skipWhitespace()
                } else if char == "(" {
                    try skipSubtree() // Skip variations
                    skipWhitespace()
                } else if char == ")" {
                    advance() // Skip ')'
                    break
                } else {
                    advance() // Skip unexpected characters
                }
            }
        }

        func parseNode() throws -> SGFNode {
            var properties: [String: [String]] = [:]
            skipWhitespace()

            while let char = peek(), char.isLetter {
                let key = parseIdentifier()
                var values: [String] = []
                skipWhitespace()

                while peek() == "[" {
                    values.append(parseValue())
                    skipWhitespace()
                }

                if !values.isEmpty {
                    properties[key, default: []].append(contentsOf: values)
                }
                skipWhitespace()
            }

            return SGFNode(properties: properties)
        }

        func parseIdentifier() -> String {
            let start = index
            while let char = peek(), char.isLetter {
                advance()
            }
            return String(normalizedText[start..<index]).uppercased()
        }

        func parseValue() -> String {
            guard peek() == "[" else {
                return ""
            }
            advance() // Skip '['

            var result = ""
            while let char = peek() {
                advance()
                if char == "\\" {
                    if let nextChar = peek() {
                        result.append(nextChar)
                        advance()
                    }
                } else if char == "]" {
                    break
                } else {
                    result.append(char)
                }
            }
            return result
        }

        func skipSubtree() throws {
            guard peek() == "(" else {
                throw SGFParsingError.parseError("Expected '(' for subtree")
            }
            advance() // Skip '('

            var depth = 1
            while depth > 0 && peek() != nil {
                let char = peek()!
                advance()
                if char == "(" {
                    depth += 1
                } else if char == ")" {
                    depth -= 1
                }
            }
        }

        // Start parsing
        skipWhitespace()
        guard peek() == "(" else {
            throw SGFParsingError.parseError("SGF must start with '('")
        }

        try parseSubtree()
        return SGFTree(nodes: nodes)
    }

    /// Parse SGF text directly into a Game object
    public static func parseGame(_ text: String) throws -> Game {
        let tree = try parseTree(text)
        return try convertTreeToGame(tree)
    }

    /// Convert parsed SGF tree into Game object
    public static func convertTreeToGame(_ tree: SGFTree) throws -> Game {
        var gameInfo = GameInfo()
        var boardSize = 19
        var setup: [PlacedStone] = []
        var moves = MoveSequence()

        for node in tree.nodes {
            // Parse game info
            if let event = node.getValue("EV") {
                gameInfo.event = event
            }
            if let blackPlayer = node.getValue("PB") {
                gameInfo.playerBlack = blackPlayer
            }
            if let whitePlayer = node.getValue("PW") {
                gameInfo.playerWhite = whitePlayer
            }
            if let result = node.getValue("RE") {
                gameInfo.result = result
            }
            if let date = node.getValue("DT") {
                gameInfo.date = date
            }
            if let rules = node.getValue("RU") {
                gameInfo.rules = rules
            }
            if let komiStr = node.getValue("KM"), let komi = Double(komiStr) {
                gameInfo.komi = komi
            }
            if let handicapStr = node.getValue("HA"), let handicap = Int(handicapStr) {
                gameInfo.handicap = handicap
            }

            // Parse board size
            if let sizeStr = node.getValue("SZ") {
                if let colonIndex = sizeStr.firstIndex(of: ":") {
                    let leftStr = String(sizeStr[..<colonIndex])
                    let rightStr = String(sizeStr[sizeStr.index(after: colonIndex)...])
                    let left = Int(leftStr) ?? 19
                    let right = Int(rightStr) ?? left
                    boardSize = max(2, min(25, min(left, right)))
                } else if let size = Int(sizeStr) {
                    boardSize = max(2, min(25, size))
                }
            }

            // Parse setup stones
            for blackSetup in node.getValues("AB") {
                if let position = Position(sgfCoordinate: blackSetup) {
                    setup.append(PlacedStone(stone: .black, position: position))
                }
            }
            for whiteSetup in node.getValues("AW") {
                if let position = Position(sgfCoordinate: whiteSetup) {
                    setup.append(PlacedStone(stone: .white, position: position))
                }
            }

            // Parse moves
            if let blackMove = node.getValue("B") {
                if blackMove.isEmpty {
                    moves.pass(.black)
                } else if let position = Position(sgfCoordinate: blackMove) {
                    moves.place(.black, at: position)
                }
            }
            if let whiteMove = node.getValue("W") {
                if whiteMove.isEmpty {
                    moves.pass(.white)
                } else if let position = Position(sgfCoordinate: whiteMove) {
                    moves.place(.white, at: position)
                }
            }
        }

        return Game(
            info: gameInfo,
            boardSize: boardSize,
            setup: setup,
            moves: moves
        )
    }
}

// MARK: - Compatibility with existing codebase

extension SGFTree {
    /// Convert to legacy format
    public var legacyNodes: [SGFNode] {
        return nodes.map { node in
            var legacyNode = SGFNode()
            legacyNode.properties = node.properties
            return legacyNode
        }
    }
}

/// Legacy SGF parsing function for compatibility
public func parseSGF(_ text: String) throws -> SGFGame {
    let game = try SGFParser.parseGame(text)
    return game.legacyGame
}